client/src/context/FetchDataContext.tsx

```diff
@@ -20,12 +29,217 @@ export const FetchDataProvider = ({ children }: { children: ReactNode }) => {
+    } catch (error) {
+      console.error("Error fetching access token:", error);
+      throw error;
+    }
+  };
+
+  const fetchMusicData = async (
+    bubbleTags: string[],
+    filter: string,
+    tags: string[]
+  ) => {
+    const tagsIfEmpty = [];
+    for (let i = 0; i < 5; i++) {
+      const randomTag = tags[Math.floor(Math.random() * tags.length)];
+      tagsIfEmpty.push(randomTag);
+    }
+
+    const randomOffset = Math.floor(Math.random() * 1000);
+    const API_URL =
+      bubbleTags.length === 0
+        ? `https://api.spotify.com/v1/search?q="${tagsIfEmpty.join(
+            "+"
+          )}"&type=track&limit=1&offset=${randomOffset}`
+        : `https://api.spotify.com/v1/search?q="${bubbleTags.join(
+            "+"
+          )}"&type=track&limit=1&offset=${randomOffset}`;
+
+    setLoading(true);
+    setError(null);
+
+    try {
+      const token = await getAccessToken();
+
+      const response = await fetch(API_URL, {
+        method: "GET",
+        headers: {
+          Authorization: `Bearer ${token}`,
+          "Content-Type": "application/json",
+        },
+      });
+
+      if (!response.ok) {
+        throw new Error(`HTTP error! status: ${response.status}`);
+      }
+
+      const result = await response.json();
+      if (
+        result.tracks &&
+        Array.isArray(result.tracks.items) &&
+        result.tracks.items.length > 0
+      ) {
+        const item = result.tracks.items[0];
+        const popularity = item.popularity;
+
+        const isValid = () => {
+          if (filter === "Unknown") return popularity >= 0 && popularity <= 5;
+          if (filter === "Low") return popularity >= 6 && popularity <= 15;
+          if (filter === "Medium") return popularity >= 16 && popularity <= 30;
+          return true;
+        };
+
+        if (
+          musicHistory.some(
+            (music) => music.name === item.name || music.id === item.id
+          )
+        ) {
+          console.log("Duplicate item found, fetching again...");
+          return fetchMusicData(bubbleTags, filter, tags);
+        }
+
+        if (!isValid()) {
+          return fetchMusicData(bubbleTags, filter, tags);
+        }
+
+        const artistId = item.artists[0]?.id;
+        if (!artistId) {
+          throw new Error("Artist ID not found");
+        }
+
+        const artistData = await fetchArtistData(artistId, token);
+        const albumsData = await fetchAlbumsData(artistId, token);
+
+        const newSelectedMusic = new SelectedMusicModel(
+          item.id,
+          item.popularity,
+          item.external_urls.spotify,
+          item.name,
+          item.artists[0]?.name,
+          item.album.images[0]?.url || "",
+          artistData.artistImage,
+          artistData.genre,
+          artistData.followers,
+          albumsData
+        );
+
+        musicHistory.push(newSelectedMusic);
+        setMusicHistory(musicHistory);
+        console.log("Music history:", musicHistory);
```

Comment:
there are a lotr of console logs in a lot of files, please remove them before pushing again